packages/snaps-rpc-methods/src/permitted/getState.ts

```diff
@@ -0,0 +1,189 @@
+
+import { manageStateBuilder } from '../restricted/manageState';
+import type { MethodHooksObject } from '../utils';
+import { StateKeyStruct } from '../utils';
+
+const hookNames: MethodHooksObject<GetStateHooks> = {
+  hasPermission: true,
+  getSnapState: true,
+  getUnlockPromise: true,
+};
+
+/**
+ * `snap_getState` gets the state of the Snap.
+ */
+export const getStateHandler: PermittedHandlerExport<
+  GetStateHooks,
+  GetStateParameters,
+  GetStateResult
+> = {
+  methodNames: ['snap_getState'],
+  implementation: getStateImplementation,
+  hookNames,
+};
+
+export type GetStateHooks = {
+  /**
+   * Check if the requesting origin has a given permission.
+   *
+   * @param permissionName - The name of the permission to check.
+   * @returns Whether the origin has the permission.
+   */
+  hasPermission: (permissionName: string) => boolean;
+
+  /**
+   * Get the state of the requesting Snap.
+   *
+   * @returns The current state of the Snap.
+   */
+  getSnapState: (
+    snapId: string,
+    encrypted: boolean,
+  ) => Promise<Record<string, Json>>;
+
+  /**
+   * Wait for the extension to be unlocked.
+   *
+   * @returns A promise that resolves once the extension is unlocked.
+   */
+  getUnlockPromise: (shouldShowUnlockRequest: boolean) => Promise<void>;
+};
+
+const GetStateParametersStruct = object({
+  key: optional(StateKeyStruct),
+  encrypted: optional(boolean()),
+});
+
+export type GetStateParameters = InferMatching<
+  typeof GetStateParametersStruct,
+  GetStateParams
+>;
+
+/**
+ * The `snap_getState` method implementation.
+ *
+ * @param request - The JSON-RPC request object.
+ * @param response - The JSON-RPC response object.
+ * @param _next - The `json-rpc-engine` "next" callback. Not used by this
+ * function.
+ * @param end - The `json-rpc-engine` "end" callback.
+ * @param hooks - The RPC method hooks.
+ * @param hooks.hasPermission - Check whether a given origin has a given
+ * permission.
+ * @param hooks.getSnapState - Get the state of the requesting Snap.
+ * @param hooks.getUnlockPromise - Wait for the extension to be unlocked.
+ * @returns Nothing.
+ */
+async function getStateImplementation(
+  request: JsonRpcRequest<GetStateParameters>,
+  response: PendingJsonRpcResponse<GetStateResult>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  { hasPermission, getSnapState, getUnlockPromise }: GetStateHooks,
+): Promise<void> {
+  const { params } = request;
+
+  if (!hasPermission(manageStateBuilder.targetName)) {
+    return end(providerErrors.unauthorized());
+  }
+
+  try {
+    const validatedParams = getValidatedParams(params);
+    const { key, encrypted = true } = validatedParams;
+
+    if (encrypted) {
+      await getUnlockPromise(true);
+    }
+
+    // We expect the MM middleware stack to always add the origin to requests
+    const { origin } = request as JsonRpcRequest & { origin: string };
+    const state = await getSnapState(origin, encrypted);
```

Comment:
It would be awesome if we managed to squeeze out more performance by doing this, e.g. by only having to decrypt part of the state blob instead of the entire thing.

That'll be impossible to do without breaking changes or writing some migration code, but perhaps this is the time to consider it. Right now, this API just moves the problem that Snaps currently have onto us, which is good, but there may be more to explore here.

Reply:
I thought about it, but I'm unsure.

A developer may want to fetch multiple keys, resulting in multiple decryptions. We could cache the full state now (not sure why we don't do this already), but if we split it up, caching isn't really as useful, and the total decryption time may be longer.

It also makes it harder to receive the full state, as individual keys would be encrypted. That could be fixed by requiring a key, and not having a concept of "full state", maybe? But then depending on how the developer ends up using state, performance could be better or worse (i.e., one key for each value, or multiple values in a single key may make a difference, but I'd need to benchmark to know for sure). Unless some algorithm exists where we can encrypt a full state but only decrypt parts of it?

Have we considered caching decrypted state?

Reply:
> Have we considered caching decrypted state?

I can't recall if we've discussed it, but I have thought about myself. I think we would have to consider the security implications of doing that (e.g. what does the vault implementation do?) as well as deciding on when to actually handle persisting the changes.